txjuju/config.py

```diff
@@ -0,0 +1,130 @@
+            BootstrapConfig(default_series, admin_secret),
+            )
+
+    def __new__(cls, name, cloud, bootstrap=None):
+        """
+        @param name: The name of the controller.
+        @param cloud: The controller's cloud config.
+        @param bootstrap: The controller's bootstrap config, if any.
+        """
+        name = unicode(name) if name else None
+        cloud = CloudConfig._make(cloud) if cloud else None
+        if bootstrap is None:
+            bootstrap = BootstrapConfig("")
+        else:
+            bootstrap = BootstrapConfig._make(bootstrap)
+        return super(ControllerConfig, cls).__new__(
+            cls, name, cloud, bootstrap)
+
+    def __init__(self, *args, **kwargs):
+        super(ControllerConfig, self).__init__(*args, **kwargs)
+        if not self.name:
+            raise ValueError("missing name")
+        if not self.cloud:
+            raise ValueError("missing cloud")
+        if not self.bootstrap:
+            raise ValueError("missing bootstrap")
+
+
+class CloudConfig(
+        namedtuple("CloudConfig",
+                   "name type endpoint auth_types credentials")):
+    """An encapsulation of the local config for a single cloud."""
+
+    def __new__(cls, name, type=None, endpoint=None,
+                auth_types=None, credentials=None):
+        """
+        @param name: The cloud's user-defined ID.
+        @param type: The provider type (defaults to the name).
+        @param endpoint: The endpoint to use, if needed.
+        @param auth_types: The set of supported auth types, if needed.
+        @param credentials: The set of credentials configured
+            for this cloud, if any.
+        """
+        name = unicode(name) if name else None
+        if type is None:
+            type = name
+        type = unicode(type) if type else None
+        endpoint = unicode(endpoint) if endpoint else None
+        # TODO Add provider-specific abstractions to support auth_types?
+        # TODO Add support for auth_types and credentials as soon as needed.
+        auth_types = tuple(auth_types) if auth_types else None
+        credentials = tuple(credentials) if credentials else None
+        return super(CloudConfig, cls).__new__(
+            cls, name, type, endpoint, auth_types, credentials)
+
+    def __init__(self, *args, **kwargs):
+        super(CloudConfig, self).__init__(*args, **kwargs)
+        if not self.name:
+            raise ValueError("missing name")
+        if not self.type:
+            raise ValueError("missing type")
+        # TODO Add support for auth_types and credentials as soon as needed.
+        if self.auth_types is not None or self.credentials is not None:
+            raise NotImplementedError
+
+
+# TODO Implement CloudCredentialConfig as soon as needed.
+
+
+class BootstrapConfig(
+        namedtuple("BootstrapConfig", "default_series admin_secret")):
+    """An encapsulation of a bootstrap config."""
```

Comment:
My initial focus is on fake-juju.  We can circle back on this when we update landscape to use txjuju.config.